deep_qa/common/params.py

```diff
@@ -140,47 +169,31 @@ def __iter__(self):
     def __len__(self):
         return len(self.params)
 
-
-def _get_choice_error_message(value: Any, choices: List[Any], name: str=None) -> str:
-    if name:
-        return '%s not in acceptable choices for %s: %s' % (value, name, str(choices))
-    else:
-        return '%s not in acceptable choices: %s' % (value, str(choices))
-
-
-def pop_choice_with_default(params: Dict,
-                            key: str,
-                            choices: List[Any],
-                            default: Any=None,
-                            name: str=None) -> Any:
-    """
-    Performs the same function as the :func:`Params.pop_with_default_method` of Params,
-    but is required in order to deal with places that the Params object is not
-    welcome, such as inside Keras layers.
-    """
-    try:
-        value = params.pop(key)
-    except KeyError:
-        if default is None:
-            value = choices[0]
+    def __check_is_dict(self, new_history, value):
+        if isinstance(value, dict):
+            new_history = self.history + new_history + "."
+            return Params(value, new_history)
         else:
-            value = default
-    if value not in choices:
-        raise ConfigurationError(_get_choice_error_message(default, choices, name))
-
-    logger.param("UNSCOPED PARAMETER: Retrieve default arguments manually."
-                 + key + " : " + str(value))
-
-    return value
+            return value
 
 
-def assert_params_empty(params: Union[Params, Dict], class_name: str):
+def pop_choice(params: Dict[str, Any],
+               key: str,
+               choices: List[Any],
+               default_to_first_choice: bool=False,
+               history: str="?.") -> Any:
     """
-    Raises a ConfigurationError if ``params`` is not empty, with a message about where the extra
-    parameters were passed to.
+    Performs the same function as :func:`Params.pop_choice`, but is required in order to deal with
+    places that the Params object is not welcome, such as inside Keras layers.  See the docstring
+    of that method for more detail on how this function works.
+
+    This method adds a ``history`` parameter, in the off-chance that you know it, so that we can
+    reproduce :func:`Params.pop_choice` exactly.  We default to using "?." if you don't know the
+    history, so you'll have to fix that in the log if you want to actually recover the logged
+    parameters.
     """
-    if len(params) != 0:
-        raise ConfigurationError("Extra parameters passed to {}: {}".format(class_name, params))
+    value = Params(params, history).pop_choice(key, choices, default_to_first_choice)
+    return value
```

Comment:
It only serializes stuff that shows up in `get_config`.  So we save all of the layer parameters to `self` as we got them (as `dicts`), and return those in `get_config`, so it doesn't matter what we do with them after that.